mobilecoind/src/service.rs

```diff
@@ -505,6 +506,83 @@ impl<T: BlockchainConnection + UserTxConnection + 'static, FPR: FogPubkeyResolve
+            .map_err(|err| rpc_invalid_arg_error("PublicAddress.try_from", err, &self.logger))?;
+
+        let hash = ShortAddressHash::from(&address);
+
+        let mut response = api::GetShortAddressHashResponse::new();
+        response.set_hash(hash.as_ref().to_vec());
+        Ok(response)
+    }
+
+    fn validate_authenticated_sender_memo_impl(
+        &mut self,
+        request: api::ValidateAuthenticatedSenderMemoRequest,
+    ) -> Result<api::ValidateAuthenticatedSenderMemoResponse, RpcStatus> {
+        // Read the utxo proto
+        let utxo = UnspentTxOut::try_from(request.get_utxo())
+            .map_err(|err| rpc_invalid_arg_error("unspent_tx_out.try_from", err, &self.logger))?;
+
+        let memo_payload = MemoPayload::try_from(&utxo.memo_payload[..])
+            .map_err(|err| rpc_invalid_arg_error("memo_payload.try_from", err, &self.logger))?;
+
+        // Read the sender proto
+        let sender = PublicAddress::try_from(request.get_sender())
+            .map_err(|err| rpc_invalid_arg_error("PublicAddress.try_from", err, &self.logger))?;
+
+        // Get MonitorId from from the GRPC request.
+        let monitor_id = MonitorId::try_from(&request.monitor_id)
+            .map_err(|err| rpc_invalid_arg_error("monitor_id.try_from.bytes", err, &self.logger))?;
+
+        // Get monitor data.
+        let data = self
+            .mobilecoind_db
+            .get_monitor_data(&monitor_id)
+            .map_err(|err| {
+                rpc_internal_error("mobilecoind_db.get_monitor_data", err, &self.logger)
+            })?;
+
+        let subaddress_vpk = data
+            .account_key
+            .subaddress_view_private(utxo.subaddress_index);
+        let tx_out_public_key = &utxo.tx_out.public_key;
+
+        let mut response = api::ValidateAuthenticatedSenderMemoResponse::new();
+
+        response.set_success(bool::from(match MemoType::try_from(&memo_payload) {
+            Ok(MemoType::AuthenticatedSender(memo)) => {
+                memo.validate(&sender, &subaddress_vpk, tx_out_public_key)
+            }
+            Ok(MemoType::AuthenticatedSenderWithPaymentRequestId(memo)) => {
+                memo.validate(&sender, &subaddress_vpk, tx_out_public_key)
+            }
+            Ok(MemoType::AuthenticatedSenderWithPaymentIntentId(memo)) => {
+                memo.validate(&sender, &subaddress_vpk, tx_out_public_key)
+            }
+            Ok(other) => {
+                return Err(rpc_invalid_arg_error(
+                    "Not an authenticated sender memo",
+                    format!("{other:?}"),
+                    &self.logger,
+                ));
+            }
+            Err(err) => {
+                return Err(rpc_invalid_arg_error(
+                    "Not an authenticated sender memo",
+                    format!("{err:?}"),
+                    &self.logger,
+                ));
+            }
+        }));
```

Comment:
I think it might make more sense for this to return a valid response with success=false, or perhaps there should be an enum in the result instead of a boolean that can specify `valid/invalid/authenticated_sender_memo_not_present`. I think the arguments in the request were technically valid, they just don't point at a valid memo.

Reply:
One thing that I've been slowly trying to improve in the mobilecoind APIs here and there is that, in many cases the service just reports "internal error", even when the request is non-sensical and cannot be satisfied even if retried. But often it's helpful to return 500-type errors when the request could be retried, and 400-type errors when something is wrong with the request and it should not be retried, because it makes it easier for the caller to handle errors appropriately.

This is a case where I feel like:
* Passing this UTXO to the `validate_authenticated_sender_memo` was a logic error in the caller, and not merely some internal error. Intuitively I want to pass back some 400-type response that indicates that this error is not retriable, because I can easily match on this at top level when doing error handling. Invalid arg error seemed like a decent option but I'm open to others.
* There are other cases like, when the request takes `bytes monitor_id`. In that case, we always have to do a conversion to see if there are exactly 32 bytes, and if not then it's not a valid monitor id, and we are returning `invalid_arg_error`. Do you also think that's bad? Even if the request is a valid protobuf, it's still nonsensical because those bytes are not a valid monitor id, right? To me these two situations feel similar. Even if the arguments in this `validate_authenticated_sender_memo` are technically valid protobuf, if it's not an authenticated sender memo, it seems reasonable to say this is outside the preconditions of the call, and return an invalid arg error.

Reply:
happy to go whatever way you guys want, that's just my 2 cents

Reply:
I think `invalid_arg_error` makes sense for the `monitor_id` example, since the data is in fact invalid.
In the UTXO-without-authenticated-sender-memo case, I think the UTXO is in fact valid, it just happens that it does not contain an authenticated sender memo. I am not sure passing it is a logic error from the caller. For example, I can see a scenario where caller only wants to process UTXOs that have valid authenticated sender memos - with the invalid_arg implementation they have to take two steps (first to see if the UTXO has an authenticated sender memo, and then check if it is valid), versus just checking if its valid and proceeding based on the response.

Reply:
So, I can see how that pattern might make sense for some RPC's, but for this one, the way it's intended to be used is like:

* I got a UTXO, and I want to try to verify the sender using RTH
* I check if it has a sender memo, and read the hash out of that (if there is no memo, then we simply can't verify the sender this way, and we have to give up)
* I go look up the public address based on the hash
* I supply that public address, and this UTXO, to the `validate_authenticated_sender_memo` RPC call, to check the hmac

So in this pattern, you already have to check if there's a sender memo, because you have to read the hash out of it.

Conceivably someone might write code that uses this in a blind way, where, they test public addresses against UTXOs without checking the hash first. But it's to understand why they would do that, it would just be very inefficient because checking the hash is very inexpensive compared to making an RPC call, and if you have a lot of public addresses it would be a lot of RPC calls.

But the other thing I would say is:

>  - with the invalid_arg implementation they have to take two steps (first to see if the UTXO has an authenticated sender memo, and then check if it is valid), versus just checking if its valid and proceeding based on the response.

I think they still could write the same code, they just have to check for the invalid_arg error code if they want to use the function that way.

> I think the UTXO is in fact valid, it just happens that it does not contain an authenticated sender memo.

I guess in my thinking, data can be valid or invalid depending on context. But I can see that there is also value to reserving the "invalid arg" errors to strictly mean "we had some kind of deserialization issue", so maybe it's useful to only use that code in that case. And here we don't have a deserialization issue.

Reply:
grpc docu seems to suggest: https://grpc.github.io/grpc/core/md_doc_statuscodes.html

```
INVALID_ARGUMENT 	3 	The client specified an invalid argument. Note that this differs from FAILED_PRECONDITION. INVALID_ARGUMENT indicates arguments that are problematic regardless of the state of the system (e.g., a malformed file name).
```

Reply:
I don't feel strongly about this, in reality it doesn't matter that much so either way is acceptable.